src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1713,11 +1713,45 @@ abstract class RefChecks extends Transform {
         if (!inPattern) {
           checkImplicitViewOptionApply(tree.pos, fn, args)
           checkSensible(tree.pos, fn, args) // TODO: this should move to preEraseApply, as reasoning about runtime semantics makes more sense in the JVM type system
+          checkNamedBooleanArgs(fn, args)
         }
         currentApplication = tree
         tree
     }
 
+    /** Check that boolean literals are passed as named args.
+     *  The rule is enforced when the type of the parameter is `Boolean`.
+     *  The rule is relaxed when the method has exactly one boolean parameter
+     *  and it is the first parameter, such as `assert(false, msg)`.
+     */
+    private def checkNamedBooleanArgs(fn: Tree, args: List[Tree]): Unit = {
+      val sym = fn.symbol
+      def applyDepth: Int = {
+        def loop(t: Tree, d: Int): Int =
+          t match {
+            case Apply(f, _) => loop(f, d+1)
+            case _ => d
+          }
+        loop(fn, 0)
+      }
+      def isAssertParadigm(params: List[Symbol]): Boolean = !sym.isConstructor && !sym.isCaseApplyOrUnapply && {
```

Comment:
I'm not convinced we should make a difference between constructors and other methods here.

Reply:
classes are less likely to have a name that makes its parameter meaningful, unlike a method which is of narrower semantic scope.